barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -509,4 +514,95 @@ namespace avm_alu(256);
+    //
+    // === Steps when performing SHL
+    // (1) Prove the correct decomposition: a_hi * 2**(t-b) + a_lo = a
+    // (2) Range check a_hi < 2**b && a_lo < 2**(t-b)
+    // (3) Return a_lo * 2**b
+    // 
+    //  <-- b bits -->   | <-- (t-b) bits -->
+    // ------------------|-------------------
+    // |      a_hi       |      a_lo        | --> a
+    // --------------------------------------
+
+    // TODO: Possibly optimised with variable length checks
+    // Check that a_lo and a_hi are range checked such that that:
+    // SHR: a_hi < 2**(t - b) && a_lo < 2**b
+    // SHL: a_hi < 2**b && a_lo < 2**(t - b)
+
+    // In lieu of a variable length check, we can utilise 2 fixed range checks instead.
+    // Given the dynamic range check of 0 <= a_hi <= 2**(t-b), where b < t
+    // (1) 0 <= a_hi <= 2**t
+    // (2) 0 <= 2**(t - b)  - a_hi <= 2**t
+    // Note that (1) is guaranteed elsewhere through the tagged memory model, so we focus on (2) here. 
+
+    // === General Notes:
+    // There are probably ways to merge various relations for the SHL/SHR, but they are separate
+    // now while we are still figuring out.
+
+    // Indicate if the shift amount < MAX_BITS
+    pol commit shift_lt_bit_len;
+    shift_lt_bit_len * (1 - shift_lt_bit_len) = 0;
+
+    // The number of bits represented by the memory tag, any shifts greater than this will result in zero.
+    pol MAX_BITS = u8_tag * 8 + 
+                u16_tag * 16 +
+                u32_tag * 32 +
+                u64_tag * 64 +
+                u128_tag * 128;
+
+    // The result of MAX_BITS - ib
+    pol commit t_sub_b_bits;
+
+    // For our assumptions to hold, we must check that b < MAX_BITS. This can be achieved by the following relation.
+    // We check if b < MAX_BITS || b >= MAX_BITS using boolean shift_le_bit_len. 
+    pol SHIFT_DIFF = shift_lt_bit_len * (MAX_BITS - ib - 1) + (1 - shift_lt_bit_len) * (ib - MAX_BITS);
+
+    // Regardless of which side is evaluated, the value of t_sub_b_bits < 2**8
+    // so it is automatically range checked by the lookup for 2**t_sub_b_bits
+    #[SHIFT_LT_BIT_LEN]
+    t_sub_b_bits = shift_sel * (SHIFT_DIFF + shift_lt_bit_len); // Re-add the 1
+
+    // Lookups for powers of 2. 
+    // 2**(MAX_BITS - ib)
+    pol commit two_pow_t_sub_b;
+    // 2 ** ib
+    pol commit two_pow_b; 
+
+    // ========= SHIFT RIGHT OPERATIONS ===============================
+    // a_hi * 2**b + a_lo = a
+    // If ib > MAX_BITS, we trivially skip this check since the result will be forced to 0.
+    #[CHECK_INPUT_DECOMPOSITION_0]
+    shift_lt_bit_len * op_shr * ((b_hi * two_pow_b + b_lo) - ia) = 0;
+
+    // Return hi limb, if ib > MAX_BITS, the output is forced to be 0
+    #[SHR_OUTPUT_0]
+    shift_lt_bit_len * op_shr * (b_hi - ic) = 0;
+
+    // ========= SHIFT LEFT OPERATIONS ===============================
+    // a_hi * 2**(t-b) + a_lo = a
+    // If ib > MAX_BITS, we trivially skip this check since the result will be forced to 0.
```

Comment:
I think we are missing a relation like:
(1 - shift_lt_bit_len) *(op_shl + op_shr) * ic = 0;

Reply:
yep you're right, ill try and see if i can rephrase the existing return relations

Reply:
So i've reworked the return relations, i tink we just needed to switch the paranthesis ordering. I believe this stops ic from being a free variable if shift_lt_bit_len is set

```
#[SHL_OUTPUT_0] --> old
shift_lt_bit_len * op_shr * (b_hi - ic) = 0;

#[SHL_OUTPUT_0] --> new
op_shr * (ic - (b_hi * shift_lt_bit_len)) = 0;
```
and
```
#[SHL_OUTPUT_1] --> old
shift_lt_bit_len * (op_shl * (b_lo * two_pow_b - ic)) = 0;

#[SHL_OUTPUT_1] --> new
op_shl * (ic - (b_lo * two_pow_b * shift_lt_bit_len)) = 0;
```

Reply:
Excellent idea!